File: frontend/src/types/price-sheets.ts
import type { ApiErrorBody } from './offers'

export type PriceSheetBatchStatus =
  | 'reviewing' | 'queued' | 'running' | 'paused' | 'waiting_user'
  | 'completed' | 'completed_partial' | 'stopped' | 'failed'

export interface PriceSheetBatchView {
  id: number
  file_name: string
  price_date: string
  date_inferred: boolean
  status: PriceSheetBatchStatus
  recognized_count: number
  selected_count: number
  completed_item_count: number
  partial_item_count: number
  failed_item_count: number
  lower_price_count: number
  current_item_id: number | null
  pause_requested: boolean
  stop_requested: boolean
  last_error_code: string | null
  last_error_summary: string | null
  created_at: string
  updated_at: string
  started_at: string | null
  finished_at: string | null
}

export interface PriceSheetItemView {
  id: number
  batch_id: number
  sequence: number
  selected: boolean
  brand: string
  model_name: string
  storage: string
  color: string
  today_price_cents: number
  raw_text: string
  confidence: number
  review_required: boolean
  status: string
  candidate_count: number
  total_region_count: number
  completed_region_count: number
  failed_region_count: number
  lowest_price_cents: number | null
  last_error_code: string | null
  last_error_summary: string | null
  started_at: string | null
  finished_at: string | null
}

export interface PriceSheetRegionTaskView {
  id: number
  price_sheet_item_id: number
  region_code: string
  province: string
  city: string
  district: string
  street: string
  sequence: number
  status: string
  attempts: number
  verified_candidate_count: number
  lowest_result_cents: number | null
  error_code: string | null
  error_summary: string | null
  started_at: string | null
  finished_at: string | null
}

export interface PriceSheetCheckoutCurrentView {
  platform_sku_id: string
  region_code: string
  address: string
  entry_mode: string | null
}

export interface PriceSheetCheckoutProgressView {
  stage: string
  candidate_count: number
  task_total: number
  task_finished: number
  verified_count: number
  conditional_count: number
  address_required_count: number
  unavailable_count: number
  failed_count: number
  skipped_count: number
  cart_attention_required: boolean
  current: PriceSheetCheckoutCurrentView | null
}

export interface PriceSheetBatchDetail {
  batch: PriceSheetBatchView
  items: PriceSheetItemView[]
  tasks: PriceSheetRegionTaskView[]
  checkout_progress: PriceSheetCheckoutProgressView
}

export interface PriceSheetResultView {
  item_id: number
  model_name: string
  storage: string
  color: string
  today_price_cents: number
  status: string
  coverage: string
  region_code: string | null
  address: string | null
  platform_sku_id: string | null
  title: string | null
  product_url: string | null
  shop_name: string | null
  entry_mode: string | null
  price_status: string | null
  quantity: number | null
  target_only: boolean | null
  line_original_price_cents: number | null
  line_sale_price_cents: number | null
  merchant_discount_cents: number | null
  ordinary_coupon_cents: number | null
  subsidy_amount_cents: number | null
  shipping_fee_cents: number | null
  payable_price_cents: number | null
  discount_summary: string | null
  conditional_reason: string | null
  cart_restored: boolean | null
  failed_count: number
  captured_at: string | null
}

export interface PriceSheetResultsView {
  lower_results: PriceSheetResultView[]
  not_lower_items: PriceSheetResultView[]
  partial_items: PriceSheetResultView[]
}

export type PriceSheetError = ApiErrorBody

File: backend/app/schemas/price_sheets.py
from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator


class PriceSheetItemInput(BaseModel):
    selected: bool = True
    brand: str = Field(min_length=1, max_length=80)
    model_name: str = Field(min_length=1, max_length=160)
    storage: str = Field(min_length=1, max_length=40)
    color: str = Field(min_length=1, max_length=80)
    today_price_cents: int = Field(ge=100_000, le=3_000_000)
    raw_text: str = Field(default="", max_length=2000)
    confidence: float = Field(ge=0, le=1)
    review_required: bool = False

    @field_validator("brand", "model_name", "storage", "color")
    @classmethod
    def strip_required_text(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if not normalized:
            raise ValueError("字段不能为空")
        return normalized


class PriceSheetItemsUpdate(BaseModel):
    price_date: date
    items: list[PriceSheetItemInput] = Field(max_length=200)


class PriceSheetBatchView(BaseModel):
    id: int
    file_name: str
    price_date: date
    date_inferred: bool
    status: str
    recognized_count: int
    selected_count: int
    completed_item_count: int
    partial_item_count: int
    failed_item_count: int
    lower_price_count: int
    current_item_id: int | None
    pause_requested: bool
    stop_requested: bool
    last_error_code: str | None
    last_error_summary: str | None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    finished_at: datetime | None


class PriceSheetItemView(PriceSheetItemInput):
    id: int
    batch_id: int
    sequence: int
    status: str
    candidate_count: int
    total_region_count: int
    completed_region_count: int
    failed_region_count: int
    lowest_price_cents: int | None
    last_error_code: str | None
    last_error_summary: str | None
    started_at: datetime | None
    finished_at: datetime | None


class PriceSheetRegionTaskView(BaseModel):
    id: int
    price_sheet_item_id: int
    region_code: str
    province: str
    city: str
    district: str
    street: str
    sequence: int
    status: str
    attempts: int
    verified_candidate_count: int
    lowest_result_cents: int | None
    error_code: str | None
    error_summary: str | None
    started_at: datetime | None
    finished_at: datetime | None


class PriceSheetCheckoutCurrentView(BaseModel):
    platform_sku_id: str
    region_code: str
    address: str
    entry_mode: str | None


class PriceSheetCheckoutProgressView(BaseModel):
    stage: str
    candidate_count: int
    task_total: int
    task_finished: int
    verified_count: int
    conditional_count: int
    address_required_count: int
    unavailable_count: int
    failed_count: int
    skipped_count: int
    cart_attention_required: bool
    current: PriceSheetCheckoutCurrentView | None = None


class PriceSheetBatchDetail(BaseModel):
    batch: PriceSheetBatchView
    items: list[PriceSheetItemView]
    tasks: list[PriceSheetRegionTaskView] = Field(default_factory=list)
    checkout_progress: PriceSheetCheckoutProgressView


class PriceSheetResultView(BaseModel):
    item_id: int
    model_name: str
    storage: str
    color: str
    today_price_cents: int
    status: str
    coverage: str
    region_code: str | None = None
    address: str | None = None
    platform_sku_id: str | None = None
    title: str | None = None
    product_url: str | None = None
    shop_name: str | None = None
    entry_mode: str | None = None
    price_status: str | None = None
    quantity: int | None = None
    target_only: bool | None = None
    line_original_price_cents: int | None = None
    line_sale_price_cents: int | None = None
    merchant_discount_cents: int | None = None
    ordinary_coupon_cents: int | None = None
    subsidy_amount_cents: int | None = None
    shipping_fee_cents: int | None = None
    payable_price_cents: int | None = None
    discount_summary: str | None = None
    conditional_reason: str | None = None
    cart_restored: bool | None = None
    failed_count: int = 0
    captured_at: datetime | None = None


class PriceSheetResultsView(BaseModel):
    lower_results: list[PriceSheetResultView]
    not_lower_items: list[PriceSheetResultView]
    partial_items: list[PriceSheetResultView]
